File: 242-valid-anagram/valid-anagram.dart
class Solution {
  bool isAnagram(String s, String t) {
    if (s.length != t.length) return false;

    final counts = Map<int, int>();
    for (final codePoint in s.runes) {
      counts[codePoint] = (counts[codePoint] ?? 0) + 1;
    }

    for (final codePoint in t.runes) {
      if (!counts.containsKey(codePoint) || counts[codePoint] == 0) {
        return false;
      }
      counts[codePoint] = counts[codePoint]! - 1;
    }

    return true;
  }
}

File: 242-valid-anagram/valid-anagram.ts
function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) return false;
  const counts: Map<string, number> = new Map();

  for (const char of s) {
    counts.set(char, (counts.get(char) ?? 0) + 1);
  }

  for (const char of t) {
    if (!counts.has(char) || counts.get(char) === 0) {
      return false;
    }
    counts.set(char, counts.get(char)! - 1);
  }

  return true;
}